### ISSUE TITLE
Error when trying to connect to mega

### ISSUE THREAD

**Issue by ChampionBuffalo1** [qgustavor/mega#85]
Hi, I'm trying to connect to my mega account using the Storage class but I keep on getting this error and I'm not sure if this has something to do with my implementation or this library so I was hoping you guys could help me with this.
Here's how I implemented it:
```js
connect = (): Promise<void> =>
    new Promise((res, rej) => {
      if (this.storage) res();
      try {
        this.storage = new Storage(
          {
            email: this.email,
            password: this.password,
          },
          (error: Error | null) => {
            // Origin: https://github.com/qgustavor/mega/blob/master/lib/storage.js#L61
            if (error) rej(error);

            logger.debug('Successfully connected to mega.');
            res();
          },
        );
      } catch (e) {
        rej(e);
      }
    });      
```
Here's the error that I'm getting
```
write EPROTO 139887624341376:error:1421C0F8:SSL routines:set_client_ciphersuite:unknown cipher returned:../deps/openssl/openssl/ssl/statem/statem_clnt.c:1335
``` 
Also, I've been using this library for a while now and I've never encountered this error before.

Node: 14.x
Npm: 7.x

**Comment by keery** [comment]
Same error

**Comment by qgustavor** [comment]
I can not reproduce this issue. Try setting `NODE_DEBUG=request` to check what was the last request before this error happened. It should be loading `https://eu.api.mega.co.nz/cs`. I just ran a test to check TLS issues using Qualys SSL Labs that resulted that there should not be any connection issue as long the client supports TLS 1.2 or higher.

Pretty much everything about how connections to the servers are handled is handled by [request](https://www.npmjs.com/package/request), which, is still pretty popular even after being deprecated. There is a plan to switch to node-fetch in future by the way. Searching `unknown cipher returned` returned many results related to proxies, so be sure to check if you are using one of those.

I will leave this issue open so people can provide more info and in case someone have an answer.

**Comment by qgustavor** [comment]
Everyone here is using Heroku? If that's the case then possible the issue is related with Heroku. Check this: https://devcenter.heroku.com/articles/troubleshooting-node-deploys#aws-proxy-error

**Comment by wiedymi** [comment]
> Everyone here is using Heroku? If that's the case then probably the issue is related with Heroku. Check this: https://devcenter.heroku.com/articles/troubleshooting-node-deploys#aws-proxy-error

@qgustavor 

This issue is reproducible locally as well. But you should wait some time to see it.

**Comment by wiedymi** [comment]
Also you should try to run node with --tls-min-v1.3 flag to see it

**Comment by datuanho2k9** [comment]
Any updates for this issue? I'm having this issue too.

**Comment by wiedymi** [comment]
> Any updates for this issue? I'm having this issue too.

Unfortunately, no. I did not find the solution. So if anyone knows how to fix it, ping me, please

**Comment by pepzwee** [comment]
@datuanho2k9 @wiedymi

According to Node docs the default minimum TLS allowed is `TLSv1.2` since version 11.4.0 _(before it used to be `TLSv1.0`)_.
Running node with `--tls-min-v1.0` should fix the issue, haven't checked if `1.1` also works.

**Comment by qgustavor** [comment]
> I am not sure why this fixed it for me especially since @qgustavor did mention that Mega does support TLS1.2.
Well, the library just calls [request](https://www.npmjs.com/package/request) which in turn, IIRC, calls native Node HTTP. What I mentioned is that Qualys SSL Labs tests shows that Mega servers require at least TLS 1.2, so if enabling 1.0 solves this issue then something is weird.

I first could not reproduce this issue, then it happened with me around three times, but it's seems pretty random and hard to replicate. I just made the code that calls the library retry if some error happens. The only bad thing about retrying is, as the current code tries to be compatible with tonistiigi's mega, it's pretty hard to handle retrying in a good way. Example: to retry a download you need call `.download()` again, which calls a API endpoint to get a download URL even if the previous download URL was still valid. Retrying login requires rehashing the password again, which is bad. I'm planning to start writing a new version of the library but I don't know how to avoid this kind of issue.

Again, about the TLS 1.0 thing, maybe Mega servers do something different with 1.0 clients, maybe it was just a coincidence, I think it needs more testing. If it solves the issue then... well, that's a good thing. Remember when you could bypass Mega download limit by opening just one download connection and clients that never bothered to implement chunked downloading were loved by people because they bypassed the limit? Mega servers are weird.

**Comment by pepzwee** [comment]
> Again, about the TLS 1.0 thing, maybe Mega servers do something different with 1.0 clients, maybe it was just a coincidence, I think it needs more testing. 

Yeah it seems to have been a coincidence. I removed the node launch option for TLS 1.0 and the same download that gave me an error before decided to work this time around. Weird.

**Comment by qgustavor** [comment]
Can you check if this issue persists in 1.0?

**Comment by ChampionBuffalo1** [comment]
It seems to be working now 😄

**Comment by ChampionBuffalo1** [comment]
There seems to be something wrong in typing, the `Storage.root.children` is an Array of `MutableFile` but in the typing files it is maked as `File[]`

**Comment by qgustavor** [comment]
That's true: when I was testing types I noticed `children` was missing in `File` but forgot `MutableFile`. I don't know a lot about TypeScript, can `MutableFile`  extend `File` replacing the definition of `children` from `File[]` to `MutableFile[]`?

**Comment by qgustavor** [comment]
I tried here and ran `ts-standard` and it worked. I will assume TypeScript allows that. I will release 1.0.1 with this fix.

**Comment by qgustavor** [comment]
Closing since the issue seems fixed.
